Feature/wx3 support

niffiwan opened the pull request.

niffiwan committed:
Upgrade to wxWidgets 3.0.2

niffiwan committed:
Add fmt module for string formating

https://github.com/fmtlib/fmt
http://fmtlib.net

niffiwan committed:
Fix wx assertions

Mostly to do with incorrect var types passed to .Format

niffiwan committed:
Standardise on use of fabs

cmath not consistently used vs math.h

niffiwan committed:
Use cmath instead of math.h

niffiwan committed:
Fix MSVC errors & warnings

niffiwan commented:
3rd attempt at the same thing from the 3rd person to attempt it? \:)

I've tested this on Linux & Windows and seems to work OK.

z64555 commented on src/Main_Panel.cpp:
```diff
@@ -187,16 +187,15 @@ main_panel::main_panel(wxFrame* parent)
 								wxPoint(widths[0]+5, 3), 
 								wxSize(widths[1], 17),
-								wxGA_PROGRESSBAR | wxGA_SMOOTH);
+								wxGA_SMOOTH);
```
Why not use wxGA_HORIZONTAL here?

niffiwan replied:
ok, done

z64555 commented on src/wxGL_PMFCanvas.h:
```diff
@@ -86,6 +86,8 @@ class wxGL_PMFCanvas : public wxGLCanvas
 
 		void draw_grid(vector3d min, vector3d max, vector3d point);
 
+		wxGLContext* m_context;
```
👍

z64555 commented on src/Main_Panel.cpp:
```diff
@@ -1192,7 +1192,11 @@ void main_panel::texture_progbar_end(wxAsyncProgressEndEvt &event)
 	UseThreadedProgBar = !skipProgdlg;
-	wxSetWorkingDirectory(wxString(filename.c_str(), wxConvUTF8).BeforeLast('\\'));
+	wxString tdir;
+	wxFileName::SplitPath(wxString(filename.c_str(), wxConvUTF8), &tdir, nullptr, nullptr);
```
could use a wxDir here?

niffiwan replied:
not sure what the benefit is to using a wxDir? Doesn't seem to be any simpler to use that when getting the dir from a filename & testing that the dir exists.
http://docs.wxwidgets.org/3.0/classwx_dir.html

z64555 commented on src/MOI.cpp:
```diff
@@ -166,7 +166,7 @@ MOI calc_under_tri(vector3d tri[3], int res){
 
 		if(s){
-			ret += calc_cuboid_MOI(vol_cent, R-L, abs(vol_cent.y*2), pz);
+			ret += calc_cuboid_MOI(vol_cent, R-L, fabs(vol_cent.y*2), pz);
```
fabs takes in a double and returns a float. Maybe fabsf would be better suited if the arguments are floats?

niffiwan replied:
ok, done, will probably have a look at all other uses of fabs as well

z64555 reviewed (commented):
Reviewed what I could. Not going to approve the merge until I can get the thing to build, though.

z64555 commented:
ILU and ILUT are currently in the "Ungrouped" category in CMake, would be good to put those under the same category as IL. Also would be good to rename the IL category to "DevIL"

TRBlount commented on src/wxGL_PMFCanvas.cpp:
```diff
@@ -395,7 +391,7 @@ void wxGL_PMFCanvas::Render()
 
 	IsRendering = true;
-	this->SetCurrent();
+	this->SetCurrent(*m_context);
```
This should go in wxGL_PMFCanvas::Init() immediately after m_context = new wxGLContext(this);

TRBlount committed:
Fix Depth Buffer Issue

niffiwan committed:
Merge pull request #1 from DahBlount/niffiwan-feature/wx3_support

Fix Depth Buffer Issue

niffiwan commented:
frak, I *thought* I'd tested having context set in Init(). 

```
../src/unix/glx11.cpp(86): assert "xid" failed in SetCurrent(): window must be shown
The program 'pcs2' received an X Window System error.
This probably reflects a bug in the program.
The error was 'GLXBadDrawable'.
  (Details: serial 820 error_code 171 request_code 154 minor_code 11)
  (Note to programmers: normally, X errors are reported asynchronously;
   that is, you will receive the error a while after causing it.
   To debug your program, run it with the --sync command line
   option to change this behavior. You can then get a meaningful
   backtrace from your debugger if you break on the gdk_x_error() function.)
```

aka X11 doesn't like your solution \:( I'll have a hunt around and see if there's another way of doing this

niffiwan committed:
Alt fix to depth buffer issue

Move all OpenGL setup to the Render function
Is this inefficient? Quite Possibly

niffiwan committed:
Cleanup opengl_init

Move to a separate function
Also remember to set default var values in ctor

niffiwan commented:
ok; thanks to [this page](https://forums.wxwidgets.org/viewtopic.php?t=29948) I've found a different way of doing this which seems to work on Linux.  Now just need to retest on my Windows laptop

niffiwan commented:
woohoo! This also worked on my Windows laptop. So if @z64555 or @DahBlount could retest & confirm this works we might have a winner (and then I can go & address @z64555's other feedback)

TRBlount commented:
opengl_init changes compile and run on my windows laptop as well.

z64555 commented:
OK, I'll give the changes a lookover as soon as I get back

z64555 commented:
Confirmed as working. Lets boogie.

niffiwan commented:
Ah. Thanks! I was intending to squash some commits & address some other comments first, but oh well \:D

niffiwan commented:
@z64555 re this comment:

> ILU and ILUT are currently in the "Ungrouped" category in CMake, would be good to put those under the same category as IL. Also would be good to rename the IL category to "DevIL"

I don't think it's possible, I tried messing around with the variable names in cmake and it breaks builds, probably because `find_package` needs to look for specific names like `IL_LIBRARIES`, `IL_INCLUDE_DIR`, etc.

edit: also grouping seems to occur only by having the same `STRING_` occuring at the start of at least two variables.
https://cmake.org/pipermail/cmake/2009-June/030372.html
